rskj-core/src/main/java/co/rsk/rpc/modules/eth/eip712/EIP712Utils.java

```diff
@@ -0,0 +1,307 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.io.ByteArrayDataOutput;
+import com.google.common.io.ByteStreams;
+import org.bouncycastle.util.encoders.Hex;
+import org.ethereum.crypto.HashUtil;
+import org.web3j.abi.DefaultFunctionEncoder;
+import org.web3j.abi.TypeDecoder;
+import org.web3j.abi.TypeReference;
+import org.web3j.abi.datatypes.NumericType;
+import org.web3j.abi.datatypes.Type;
+import org.web3j.abi.datatypes.generated.Bytes32;
+
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Spliterator;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * Encoding utilities for EIP712.
+ *
+ * Only EIP v4 implemented. Adding v3 and v1 are trivial, particularly
+ * v3, which iboild down to v4 without array handling capabilities.
+ *
+ * @see <a href="https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md">EIP 712 specification</a>
+ * @author ppedemon
+ */
+public class EIP712Utils {
+
+    private static final String EIP712Domain = "EIP712Domain";
+
+    private static final List<String> SCHEMA_KEYS = ImmutableList.of(
+            "types", "primaryType", "domain", "message"
+    );
+
+    private static final Pattern PRIMARY_TYPE_RE = Pattern.compile("^(\\w*).*");
+
+    private static final byte[] BYTE32_ZERO = new byte[32];
+
+    /**
+     * Sanitize input.
+     *
+     * @param data typed data payload to sanitize, as a Json node
+     * @return sanitized data
+     */
+    private JsonNode sanitize(JsonNode data) {
+        ObjectMapper m = new ObjectMapper();
+        ObjectNode sanitized = m.createObjectNode();
+        for (String key : SCHEMA_KEYS) {
+            if (data.has(key)) {
+                sanitized.set(key, data.get(key));
+            }
+        }
+        if (sanitized.has("types")) {
+            ObjectNode types = sanitized.with("types");
+            if (!types.has(EIP712Domain)) {
+                types.set(EIP712Domain, m.createArrayNode());
+            }
+        }
+        return sanitized;
+    }
+
+    /**
+     * Compute the transitively closed set of types dependencies on the given primary type.
+     *
+     * @param primaryType  primary type name
+     * @param types all types declared in typed data
+     * @param results list of dependent types collected so far (used as output)
+     * @return all dependent types for primary type
+     */
+    private List<String> findTyDeps(String primaryType, JsonNode types, List<String> results) {
+        Matcher m = PRIMARY_TYPE_RE.matcher(primaryType);
+        if (!m.matches()) {
```

Comment:
Not sure about the logic, but shouldn't be better to check for the group also?
```suggestion
        if (!m.matches() || m.groupCount() < 1) {
```